DBConnect.cc

```diff
@@ -54,4 +58,310 @@ void ConnectToMongoDB() {
+    if (!client) {
+        return false;
+    }
+
+    mongoc_client_set_appname(client, "carta_backend");
+
+    query = bson_new();
+    database = mongoc_client_get_database(client, "CARTA");
+
+    collection = mongoc_client_get_collection(client, "CARTA", "preferences");
+    command = BCON_NEW("ping", BCON_INT32(1));
+    retval = mongoc_client_command_simple(client, "admin", command, NULL, &reply, &error);
+    if (!retval) {
+        fprintf(stderr, "%s\n", error.message);
+        return false;
+    }
+    str = bson_as_json(&reply, NULL);
+
+    cuserid(user);
+
+    query = bson_new();
+    BSON_APPEND_UTF8(query, "username", user);
+    cursor = mongoc_collection_find_with_opts(collection, query, NULL, NULL);
+
+    struct json_object* parsed_json;
+    struct json_object* pref_name;
+    struct json_object* pref_str;
+
+    const char* char_pref_name;
+    const char* char_pref_str;
+    const char* cstr;
+
+    bson_iter_t iter;
+    while (mongoc_cursor_next(cursor, &doc)) {
+        str = bson_as_canonical_extended_json(doc, NULL);
+        parsed_json = json_tokener_parse(str);
+        if (bson_iter_init(&iter, doc)) {
+            while (bson_iter_next(&iter)) {
+                cstr = bson_iter_key(&iter);
+                if (((strcmp(cstr, "_id")) && strcmp(cstr, "username"))) {
+                    json_object_object_get_ex(parsed_json, cstr, &pref_str);
+                    char_pref_name = cstr;
+                    char_pref_str = json_object_get_string(pref_str);
+
+                    (*(ack_message_ptr->mutable_user_preferences()))[char_pref_name] = char_pref_str;
+                }
+            }
+        }
+    }
+
+    bson_destroy(query);
+
+    return true;
+}
+
+bool SaveUserPreferencesToDB(const CARTA::SetUserPreferences& request) {
+    const char* uri_string = "mongodb://localhost:27017";
+    mongoc_uri_t* uri;
+    mongoc_client_t* client;
+    mongoc_database_t* database;
+    mongoc_collection_t* collection;
+    bson_t *command, reply, *insert, prefs, *query;
+    bson_error_t error;
+    bool retval;
+    char* str;
+    char user[50];
+    bool result = true;
+
+    mongoc_init();
+
+    uri = mongoc_uri_new_with_error(uri_string, &error);
+    if (!uri) {
+        fprintf(stderr,
+            "failed to parse URI: %s\n"
+            "error message:       %s\n",
+            uri_string, error.message);
+        return EXIT_FAILURE;
+    }
+
+    client = mongoc_client_new_from_uri(uri);
+    if (!client) {
+        return EXIT_FAILURE;
+    }
+
+    mongoc_client_set_appname(client, "carta_backend");
+
+    database = mongoc_client_get_database(client, "CARTA");
+    collection = mongoc_client_get_collection(client, "CARTA", "preferences");
+
+    command = BCON_NEW("ping", BCON_INT32(1));
+    retval = mongoc_client_command_simple(client, "admin", command, NULL, &reply, &error);
+    if (!retval) {
+        fprintf(stderr, "%s\n", error.message);
+        return EXIT_FAILURE;
+    }
```

Comment:
This is near-identical in multiple functions. Could we not factor it out into another function which we can call with the command?